### PR TITLE
Add a more efficient implementation 

### PR DESCRIPTION
**Pull request by hazimehh**
@nataliaponomareva can you please check this?

### CODE REVIEW

**Comment by hazimehh** [comment]
@googlebot I signed it!

File: tf_trees/README.md [hunk 1]
```diff
@@ -1,18 +1,47 @@
...
+This repository contains a new *tree ensemble layer* (TEL) for neural networks. The layer is differentiable so SGD can be used to train the neural network (including TEL). The layer supports conditional computation for both training and inference, i.e., when updating/evaluating a certain node in the tree, only the samples that reach that node are used in computations (this is to be contrasted with the dense computations in neural networks). We provide a low-level Tensorflow implementation along with a high-level Keras API.
 
-The layer is differentiable so SGD can be used to train the neural network (including the tree layer). The layer supports conditional computation for training and inference: when updating/evaluating a certain node in the tree, only the samples that reach that node are used in computations (this is to be contrasted with the dense computations in neural networks).
+More details to be added soon.
+
+## Installation
+The installation instructions below assume that you have Python, TensorFlow, and GCC already installed.
```
Comment (nataliaponomareva) [inline tf_trees/README.md]: So since it is a custom kernel not included in the pre-built, don't you need to rebuild tensorflow with bazel for it to be picked up?

Comment (nataliaponomareva) [reply tf_trees/README.md]: I am not sure what the steps below do so that's why i am asking

File: tf_trees/README.md [hunk 2]
```diff
@@ -1,18 +1,47 @@
...
+
+# Import the layer.
+from tel import TEL
+# The documentation of TEL can be accessed as follows
+?TEL
 
-model = Sequential()
-# Add you favorite layers here...
-tree_layer = NeuralTrees(output_logits_dim=10, trees_num=5, depth=2)
+# We will fit TEL on the Boston Housing regression dataset.
```
Comment (nataliaponomareva) [inline tf_trees/README.md]: This seems like a full blown example, can we move it out to a sep .py file. Here you can leave
model = keras.Sequential()
model.add(keras.layers.BatchNormalization())
model.add(tree_layer)
model.add(other_layers...)
etc
or something like this

File: tf_trees/neural_trees_helpers.cc [hunk 3]
```diff
@@ -67,7 +55,8 @@ void ForwardPassSingleSample(const Eigen::MatrixXf& node_weights,
                              const int depth, const float smooth_step_param,
                              Eigen::VectorXf* output,
                              std::vector<Node>* tree_nodes,
-                             std::vector<int>* reachable_leaves) {
+                             std::vector<int>* reachable_leaves,
+                             int* num_reachable_leaves, bool training_mode) {
```
Comment (nataliaponomareva) [inline tf_trees/neural_trees_helpers.cc]: move the training mode before the pointers * so all const things are first
const bool training+mode
Why do you need num_reachable_leaves if you have vector of reachable leaves (len of it is the number no?)

File: tf_trees/neural_trees_helpers.cc [hunk 4]
```diff
@@ -76,37 +65,31 @@ void ForwardPassSingleSample(const Eigen::MatrixXf& node_weights,
...
-        tree[current_index].weight_input_dot_product, smooth_step_param,
-        smooth_indicator_const_1, smooth_indicator_const_2);
-
+    float probability_left = SmoothIndicator(
+        tree[current_index].weight_input_dot_product, smooth_step_param);
+    tree[current_index].routing_left_prob = probability_left;
     // Branch left if prob_left is non zero.
     if (tree[current_index].routing_left_prob > kTolerance) {
       int left_index = 2 * current_index + 1;
       tree[left_index].root_to_node_prob =
```
Comment (nataliaponomareva) [inline tf_trees/neural_trees_helpers.cc]: const
Add a comment - assume full binary tree

File: tf_trees/neural_trees_helpers.cc [hunk 5]
```diff
@@ -139,6 +122,30 @@ void ForwardPassSingleSample(const Eigen::MatrixXf& node_weights,
     *output +=
         tree[i].root_to_node_prob * leaf_weights.col(i - first_leaf_label);
   }
+  *num_reachable_leaves = reachable_leaves->size();
+
+  if (training_mode) {
+    // Mark all the internal nodes whose descendants are reachable leaves.
+    for (auto i : *reachable_leaves) {
```
Comment (nataliaponomareva) [inline tf_trees/neural_trees_helpers.cc]: leaf

File: tf_trees/neural_trees_kernels.cc [hunk 6]
```diff
@@ -191,8 +199,9 @@ class NTComputeOutputOp : public OpKernel {
     OP_REQUIRES_OK(context,
                    context->GetAttr("output_logits_dim", &output_logits_dim_));
     OP_REQUIRES_OK(context, context->GetAttr("depth", &depth_));
-    OP_REQUIRES_OK(context,
-                   context->GetAttr("smooth_step_param", &smooth_step_param_));
+    // OP_REQUIRES_OK(context,
```
Comment (nataliaponomareva) [inline tf_trees/neural_trees_kernels.cc]: remove the commented out part

File: tf_trees/tel.py [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+  depth = op.get_attr('depth')
+  parallelize_over_samples = op.get_attr('parallelize_over_samples')
+  return [tf_trees_module.nt_compute_input_and_internal_params_gradients_op(
+      grad_loss_wrt_tree_output, op.inputs[0], op.inputs[1], op.inputs[2],
+      op.inputs[3], output_logits_dim, depth, parallelize_over_samples), None]
+
+
+class TEL(keras.layers.Layer):
+    """A custom layer containing additive decision trees.
+
```
Comment (nataliaponomareva) [inline tf_trees/tel.py]: and differentiable

File: tf_trees/tel.py [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+    in the neural net.
+
+    Attributes:
+      output_logits_dim: Dimension of the output.
+      trees_num: Number of trees in the layer.
+      depth: Depth of each tree.
+      smooth_step_param: A non-negative float. Larger values make the trees
+        more likely to hard route samples (i.e., samples reach fewer leaves).
+        Note smooth_step_param = 1/gamma, where gamma is the parameter defined
+        in the TEL paper.
```
Comment (nataliaponomareva) [inline tf_trees/tel.py]: should not we limit it to be <1 as in paper experiments?

File: tf_trees/tel.py [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+        more likely to hard route samples (i.e., samples reach fewer leaves).
+        Note smooth_step_param = 1/gamma, where gamma is the parameter defined
+        in the TEL paper. 
+      sum_outputs: Boolean. If true, the outputs of the trees will be added,
+        leading to a 2D tensor of shape=[batch_size, output_logits_dim].
+        Otherwise, the tree outputs are not added and the layer output is
+        a 2D tensor of shape=[batch_size, trees_num * output_logits_dim].
+      parallelize_over_samples: Boolean, If true, parallelizes the updates over
+        the samples in each tree. Might lead to speedups when the number of
+        trees is small (at the cost of extra memory consumption).
```
Comment (nataliaponomareva) [inline tf_trees/tel.py]: in the batch

File: tf_trees/tel.py [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+      sum_outputs=True. Otherwise, a tensor of shape=[batch_size, trees_num *
+      output_logits_dim].
+    """
+
+    def __init__(self,
+                 output_logits_dim,
+                 trees_num=1,
+                 depth=3,
+                 smooth_step_param=1.0,
+                 sum_outputs=True,
```
Comment (nataliaponomareva) [inline tf_trees/tel.py]: was not it performing dense computation with value of 1?

File: tf_trees/tel.py [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+                 parallelize_over_samples=False,
+                 split_initializer=RandomUniform(-0.01, 0.01),
+                 leaf_initializer=RandomUniform(-0.01, 0.01),
+                 split_regularizer=None,
+                 leaf_regularizer=None,
+                 **kwargs):
+      """Initializes neural trees layer."""
+      self.trees_num = trees_num
```
Comment (nataliaponomareva) [inline tf_trees/tel.py]: all the self vars are usually named with _:
e.g. self._tree_num etc

File: tf_trees/tel.py [hunk 12]
```diff
@@ -0,0 +1,190 @@
...
+        self.node_weights.append(
+            self.add_weight(
+                name='node_weights',
+                shape=[input_dim, num_internal_nodes],
+                initializer=self.split_initializer,
+                trainable=True,
+                regularizer=self.split_regularizer))
+
+        self.leaf_weights.append(
+            self.add_weight(
```
Comment (nataliaponomareva) [inline tf_trees/tel.py]: name the private methods also with _:
_add_weight

File: tf_trees/tel.py [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+                initializer=self.leaf_initializer,
+                trainable=True,
+                regularizer=self.leaf_regularizer))
+
+        self.smooth_step_param_var = self.add_weight(
+                name='smooth_step_param_var',
+                shape=[],
+                initializer =
+                tf.constant_initializer(float(self.smooth_step_param)),
+                trainable=False)
```
Comment (nataliaponomareva) [inline tf_trees/tel.py]: do you need to create a var for it, it is a const. just keep it as self._smooth_step_param ?

File: tf_trees/tel.py [hunk 14]
```diff
@@ -0,0 +1,190 @@
...
+      self.smooth_step_param = smooth_step_param
+      self.parallelize_over_samples = parallelize_over_samples
+      self.sum_outputs = sum_outputs
+      self.split_initializer = keras.initializers.get(split_initializer)
+      self.leaf_initializer = keras.initializers.get(leaf_initializer)
+      self.split_regularizer = keras.regularizers.get(split_regularizer)
+      self.leaf_regularizer = keras.regularizers.get(leaf_regularizer)
+      super(TEL, self).__init__(**kwargs)
+
+    def build(self, input_shape):
```
Comment (nataliaponomareva) [inline tf_trees/tel.py]: can you add  # @Override to show that you are overriding this main method
Also a comment on what it does (e.g. preps the variables for the tree etc)

File: tf_trees/tel.py [hunk 15]
```diff
@@ -0,0 +1,190 @@
...
+
+        self.smooth_step_param_var = self.add_weight(
+                name='smooth_step_param_var',
+                shape=[],
+                initializer =
+                tf.constant_initializer(float(self.smooth_step_param)),
+                trainable=False)
+      super(TEL, self).build(input_shape)
+
+    def call(self, inputs):
```
Comment (nataliaponomareva) [inline tf_trees/tel.py]: ditto override

File: tf_trees/tel.py [hunk 16]
```diff
@@ -0,0 +1,190 @@
...
+      tree_logits = []
+      for tree_index in range(self.trees_num):
+        tree_out, _ = tf_trees_module.nt_compute_output_op(inputs,
+                                 self.node_weights[tree_index],
+                                 self.leaf_weights[tree_index],
+                                 self.smooth_step_param_var,
+                                 self.output_logits_dim, self.depth,
+                                 self.parallelize_over_samples)
+        tree_logits.append(tree_out)
+      if self.trees_num == 1:
```
Comment (nataliaponomareva) [inline tf_trees/tel.py]: may be don't do it. imagine they dont want to sum and they tune between different number of trees and the code expects concated logits at this point. Just return tree_logits if do not accumulate

File: tf_trees/README.md (hunk 1 above)
Comment (hazimehh) [reply tf_trees/README.md]: These instructions are following https://www.tensorflow.org/guide/create_op#compile_the_op_using_your_system_compiler_tensorflow_binary_installation 

Yeah it can also be installed using bazel. Should we add installation using bazel as well?

File: tf_trees/tel.py (hunk 8 above)
Comment (hazimehh) [reply tf_trees/tel.py]: P.s. this param is 1/gamma (where gamma is defined in the paper).  Should we say "Values >=1" are recommended to exploit conditional computation? I would keep this free because it can depend on the specific setup (e.g., in case batch normalization is not used larger values might be needed).

File: tf_trees/tel.py (hunk 10 above)
Comment (hazimehh) [reply tf_trees/tel.py]: you mean for the smooth_step_param? 1 is fast in our experiments

File: tf_trees/tel.py (hunk 13 above)
Comment (hazimehh) [reply tf_trees/tel.py]: Yeah because smooth_step_param is an input variable (not parameter) in nt_compute_output_op. I changed it to a var so that we can modify smooth_step_param during training (this is not possible using parameters)

File: tf_trees/tel.py (hunk 12 above)
Comment (hazimehh) [reply tf_trees/tel.py]: add_weight is a method in the base class no?

File: tf_trees/README.md (hunk 1 above)
Comment (nataliaponomareva) [reply tf_trees/README.md]: I didn't know your way. Can you add a comment with bazel too - i think it is more common, but i might be wrong

File: tf_trees/neural_trees_helpers.cc (hunk 5 above)
Comment (nataliaponomareva) [reply tf_trees/neural_trees_helpers.cc]: Oh i meant const int leaf_id : *reachable_leaves

File: tf_trees/tel.py (hunk 10 above)
Comment (nataliaponomareva) [reply tf_trees/tel.py]: oh it is gamma i think i was confusing it with. 
But anyway - should we perform some validation on this param to make sure we won't do dense computation. Eg what is the reasonable range

File: tf_trees/tel.py (hunk 12 above)
Comment (nataliaponomareva) [reply tf_trees/tel.py]: Oh, ignore me

File: tf_trees/README.md [hunk 17]
```diff
@@ -1,18 +1,47 @@
...
+More details to be added soon.
+
+## Installation
+The installation instructions below assume that you have Python and TensorFlow already installed. Use Method 1 if you have installed TensorFlow from source. Otherwise, use Method 2.
+
+### Method 1: Compile using Bazel
+First, copy the file "BUILD" (available in the tf_trees directory) to the directory "tensorflow/core/user_ops".
+Then, run the following command:
+```bash
+bazel build --config opt //tensorflow/core/user_ops:neural_trees_ops.so
```
Comment (nataliaponomareva) [inline tf_trees/README.md]: @tanzhenyu

Comment (tanzhenyu) [reply tf_trees/README.md]: Let's say "If you have TensorFlow sources installed, you can make use of TensorFlow's build system to compile your op."

File: tf_trees/README.md [hunk 18]
```diff
@@ -1,18 +1,47 @@
...
+```
+
+### Method 2: Compile using G++
```
Comment (tanzhenyu) [inline tf_trees/README.md]: Also say Tensorflow Binary Installation.

File: tf_trees/README.md [hunk 19]
```diff
@@ -1,18 +1,47 @@
...
+Note: On OS X, add the flag "-undefined dynamic_lookup" (without quotes) to the last command above.
 
 ## Example Usage
 In Keras, the layer can be used as follows:
 ```python
-from keras.models import Sequential
-from neural_trees_layer import NeuralTrees
+from tensorflow import keras
+# If the current working directory is not tf_trees, then uncomment the following
+# lines and change "/path/to/" to the parent directory of tf_trees.
```
Comment (tanzhenyu) [inline tf_trees/README.md]: Can this be done at some init file? Seems weird to have user uncomment that

File: tf_trees/demo.py [hunk 20]
```diff
@@ -0,0 +1,30 @@
...
+# We will fit TEL on the Boston Housing regression dataset.
+# First, load the dataset.
+from keras.datasets import boston_housing
+(x_train, y_train), (x_test, y_test) = boston_housing.load_data()
+
+# Define the tree layer; here we choose 10 trees, each of depth 3.
+# Note output_logits_dim is the dimension of the tree output.
+# output_logits_dim = 1 in this case, but should be equal to the
+# number of classes if used as an output layer in a classification task.
+tree_layer = TEL(output_logits_dim=1, trees_num=10, depth=3)
```
Comment (tanzhenyu) [inline tf_trees/demo.py]: maybe say "units" instead of "output_logits_dim" to be consistent with keras.layers.Dense?

File: tf_trees/tel.py [hunk 21]
```diff
@@ -0,0 +1,195 @@
...
+      split_regularizer: A Keras regularizer for the internal (splitting) nodes.
+      leaf_regularizer: A Keras regularizer for the leaves.
+    Input shape: A tensor of shape=[batch_size, input_dim].
+    Output shape: A tensor of shape=[batch_size, output_logits_dim] if
+      sum_outputs=True. Otherwise, a tensor of shape=[batch_size, trees_num *
+      output_logits_dim].
+    """
+
+    def __init__(self,
+                 output_logits_dim,
```
Comment (tanzhenyu) [inline tf_trees/tel.py]: same, use "units" here.

File: tf_trees/demo.py (hunk 20 above)
Comment (hazimehh) [reply tf_trees/demo.py]: This is a subtle point. We provide the option "sum_outputs" to either sum all the tree outputs or return each tree output separately. This option can make the meaning of "units" ambiguous..

File: tf_trees/README.md [hunk 22]
```diff
@@ -1,18 +1,47 @@
 # The Tree Ensemble Layer
 
-This repository contains a new *tree ensemble layer* which can be used anywhere in a neural network. We provide a low-level Tensorflow implementation along with a high-level Keras API.
+This repository contains a new *tree ensemble layer* (TEL) for neural networks. The layer is differentiable so SGD can be used to train the neural network (including TEL). The layer supports conditional computation for both training and inference, i.e., when updating/evaluating a certain node in the tree, only the samples that reach that node are used in computations (this is to be contrasted with the dense computations in neural networks). We provide a low-level Tensorflow implementation along with a high-level Keras API.
```
[the remark this answers is not in the document]
Comment (hazimehh) [reply tf_trees/README.md]: I think we should keep it general because it can be used with other layers as well

**Comment by mdabros** [comment]
Hi @hazimehh. 

Very interesting with a tree ensemble layer for neural networks. Will there be a paper at some point comparing this to regular tree ensemble methods and neural networks with "standard" layers?

I apologize if this is slightly off topic for the pull request.

Best regards
Mads

**Comment by tanzhenyu** [comment]
> Zhenyu, can you take a look. How do we build a custom kernel from source and make sure it is available with bazel

Done

**Comment by hazimehh** [comment]
> Hi @hazimehh.
> 
> Very interesting with a tree ensemble layer for neural networks. Will there be a paper at some point comparing this to regular tree ensemble methods and neural networks with "standard" layers?
> 
> I apologize if this is slightly off topic for the pull request.
> 
> Best regards
> Mads

Hi Mads,

Yeah we will post the paper in the next few days. It has comparisons against gradient boosted trees and dense layers.
